networkx/convert_matrix.py

```diff
@@ -1170,83 +1171,45 @@ def to_numpy_array(
-    # value with the values given by real edges.  At the end, we convert all
-    # masked values to `nonedge`. Using masked arrays fully addresses reason 1,
-    # but for reason 2, we would still have the issue with min and max if the
-    # initial values were 0.0.  Note: an initial value of +inf is appropriate
-    # for min, while an initial value of -inf is appropriate for max. When
-    # working with sum, an initial value of zero is appropriate. Ideally then,
-    # we'd want to allow users to specify both a value for nonedges and also
-    # an initial value.  For multi(di)graphs, the choice of the initial value
-    # will, in general, depend on the combinator function---sensible defaults
-    # can be provided.
+    A = np.full((nlen, nlen), fill_value=nonedge, dtype=dtype, order=order)
+
+    # Corner cases: empty nodelist or graph without any edges
+    if nlen == 0 or G.number_of_edges() == 0:
+        return A
 
+    # Map nodes to row/col in matrix
+    idx = dict(zip(nodelist, range(nlen)))
+    if len(nodelist) < len(G):
+        G = G.subgraph(nodelist).copy()
+
+    # Collect all edge weights and reduce with `multigraph_weights`
     if G.is_multigraph():
-        # Handle MultiGraphs and MultiDiGraphs
-        A = np.full((nlen, nlen), np.nan, order=order)
-        # use numpy nan-aware operations
-        operator = {sum: np.nansum, min: np.nanmin, max: np.nanmax}
-        try:
-            op = operator[multigraph_weight]
-        except Exception as err:
-            raise ValueError("multigraph_weight must be sum, min, or max") from err
-
-        for u, v, attrs in G.edges(data=True):
-            if (u in nodeset) and (v in nodeset):
-                i, j = index[u], index[v]
-                e_weight = attrs.get(weight, 1)
-                A[i, j] = op([e_weight, A[i, j]])
-                if undirected:
-                    A[j, i] = A[i, j]
+        d = defaultdict(list)
+        for u, v, wt in G.edges(data=weight, default=1.0):
+            d[(idx[u], idx[v])].append(wt)
+        i, j = np.array(list(d.keys())).T  # indices
+        wts = [multigraph_weight(ws) for ws in d.values()]  # reduced weights
     else:
-        # Graph or DiGraph, this is much faster than above
-        A = np.full((nlen, nlen), np.nan, order=order)
-        for u, nbrdict in G.adjacency():
-            for v, d in nbrdict.items():
-                try:
-                    A[index[u], index[v]] = d.get(weight, 1)
-                except KeyError:
-                    # This occurs when there are fewer desired nodes than
-                    # there are nodes in the graph: len(nodelist) < len(G)
-                    pass
-
-    A[np.isnan(A)] = nonedge
-    A = np.asarray(A, dtype=dtype)
+        i, j, wts = [], [], []
+        for u, v, wt in G.edges(data=weight, default=1.0):
+            i.append(idx[u])
+            j.append(idx[v])
+            wts.append(wt)
+
+    # Set array values with advanced indexing
+    A[i, j] = wts
```

Comment:
Not too important: This will throw an error if the user ends up passing a string in the weight attribute, we can either type check `wts` or let the user Google the `TypeError`

Reply:
Do you mean line 1209? Whether or not this raises an exception will depend on the requested dtype - if the user requests a string dtype (knowing that the weight attributes are strings), then this should now work fine, e.g.

```python
>>> G = nx.Graph()
>>> G.add_edge(0, 1, weight="ten")
>>> G.add_edge(1, 2, weight="five")
>>> nx.to_numpy_array(G, weight="weight", dtype="U5")
array([['0.0', 'ten', '0.0'],
       ['ten', '0.0', 'five'],
       ['0.0', 'five', '0.0']], dtype='<U5')
```

I may be addressing the wrong question though!

Reply:
Ah yeah, stupid me was thinking something like:
``` python
>>> G = nx.Graph()
>>> G.add_edge(0, 1, weight="ten")
>>> G.add_edge(1, 2, weight="five")
>>> nx.to_numpy_array(G, weight="weight") # without the dtype

TypeError: ......
```
(not super used to mixing `dtype` with networkx code yet :D)